File: contracts/extraction/v1/envelope.ts
import {
  assertOnlyKeys,
  readArray,
  readBoolean,
  readEnum,
  readNonEmptyString,
  readRecord,
  readRequestId,
  readUnsignedInteger,
} from "./decoder-utils";
import type {
  EnvelopeFixture,
  ExtractionManifest,
  ProtocolEnvelope,
  ResultEnvelope,
} from "./types";
import { ERROR_CODES } from "./types";

export function decodeProtocolEnvelope(
  value: unknown,
  manifest: ExtractionManifest,
): ProtocolEnvelope {
  const source = readRecord(value, "envelope");
  const type = readNonEmptyString(source.type, "type");

  if (
    readUnsignedInteger(source.protocolVersion, "protocolVersion") !==
    manifest.protocolVersion
  ) {
    throw new Error("protocolVersion: unsupported version");
  }

  if (type === "intent") {
    assertOnlyKeys(
      source,
      ["protocolVersion", "type", "requestId", "sequence", "payload"],
      "intent",
    );
    return {
      protocolVersion: manifest.protocolVersion,
      type,
      requestId: readRequestId(source.requestId),
      sequence: readUnsignedInteger(source.sequence, "sequence"),
      payload: readRecord(source.payload, "payload"),
    };
  }

  if (type === "result") {
    assertOnlyKeys(
      source,
      ["protocolVersion", "type", "requestId", "outcome"],
      "result",
    );
    return {
      protocolVersion: manifest.protocolVersion,
      type,
      requestId: readRequestId(source.requestId),
      outcome: decodeOutcome(source.outcome, manifest),
    };
  }

  throw new Error(`type: unsupported envelope type ${type}`);
}

export function decodeEnvelopeFixture(value: unknown): EnvelopeFixture {
  const source = readRecord(value, "fixture");
  assertOnlyKeys(source, ["fixtureVersion", "cases"], "fixture");

  return {
    fixtureVersion: readUnsignedInteger(
      source.fixtureVersion,
      "fixtureVersion",
    ),
    cases: readArray(source.cases, "cases").map((entry, index) => {
      const fixtureCase = readRecord(entry, `cases[${index}]`);
      assertOnlyKeys(fixtureCase, ["name", "route", "accept", "value"], "case");
      return {
        name: readNonEmptyString(fixtureCase.name, "name"),
        route: readNonEmptyString(fixtureCase.route, "route"),
        accept: readBoolean(fixtureCase.accept, "accept"),
        value: fixtureCase.value,
      };
    }),
  };
}

function decodeOutcome(
  value: unknown,
  manifest: ExtractionManifest,
): ResultEnvelope["outcome"] {
  const source = readRecord(value, "outcome");
  const status = readNonEmptyString(source.status, "outcome.status");

  if (status === "ok") {
    assertOnlyKeys(source, ["status", "data"], "outcome");
    if (!Object.hasOwn(source, "data")) {
      throw new Error("outcome.data: required field is missing");
    }
    return { status, data: source.data };
  }
  if (status === "error") {
    assertOnlyKeys(source, ["status", "error"], "outcome");
    const error = readRecord(source.error, "outcome.error");
    assertOnlyKeys(error, ["code", "retryable"], "outcome.error");
    const code = readEnum(error.code, ERROR_CODES, "outcome.error.code");
    if (!manifest.errorCodes.includes(code)) {
      throw new Error(`outcome.error.code: unknown code ${code}`);
    }
    return {
      status,
      error: {
        code,
        retryable: readBoolean(error.retryable, "outcome.error.retryable"),
      },
    };
  }

  throw new Error(`outcome.status: unsupported status ${status}`);
}

File: apps/extraction-server/src/contracts/envelope.rs
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use super::{ContractError, ErrorCode, ExtractionManifest};

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(transparent)]
pub struct ProtocolEnvelope(WireProtocolEnvelope);

#[derive(Clone, Debug, PartialEq)]
pub struct Intent<T> {
    pub request_id: Uuid,
    pub sequence: u32,
    pub payload: T,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(tag = "type", rename_all = "lowercase", deny_unknown_fields)]
enum WireProtocolEnvelope {
    Intent {
        #[serde(rename = "protocolVersion")]
        protocol_version: u32,
        #[serde(rename = "requestId")]
        request_id: Uuid,
        sequence: u32,
        payload: Map<String, Value>,
    },
    Result {
        #[serde(rename = "protocolVersion")]
        protocol_version: u32,
        #[serde(rename = "requestId")]
        request_id: Uuid,
        outcome: Outcome,
    },
}

impl WireProtocolEnvelope {
    fn validate(&self, manifest: &ExtractionManifest) -> Result<(), ContractError> {
        let protocol_version = match self {
            Self::Intent {
                protocol_version, ..
            }
            | Self::Result {
                protocol_version, ..
            } => protocol_version,
        };
        if *protocol_version != manifest.protocol_version {
            return Err(ContractError::new("protocolVersion 不受支持"));
        }

        if let Self::Result {
            outcome: Outcome::Error { error },
            ..
        } = self
        {
            if !manifest.error_codes.contains(&error.code) {
                return Err(ContractError::new("结果包含未知错误码"));
            }
        }
        Ok(())
    }
}

impl ProtocolEnvelope {
    pub fn request_id(&self) -> Uuid {
        match &self.0 {
            WireProtocolEnvelope::Intent { request_id, .. }
            | WireProtocolEnvelope::Result { request_id, .. } => *request_id,
        }
    }

    pub fn decode_intent<T: DeserializeOwned>(&self) -> Result<Intent<T>, ContractError> {
        let WireProtocolEnvelope::Intent {
            request_id,
            sequence,
            payload,
            ..
        } = &self.0
        else {
            return Err(ContractError::new("期望 intent envelope"));
        };
        let payload = serde_json::from_value(Value::Object(payload.clone()))
            .map_err(|error| ContractError::new(format!("intent payload JSON 无效: {error}")))?;
        Ok(Intent {
            request_id: *request_id,
            sequence: *sequence,
            payload,
        })
    }

    pub fn ok<T: Serialize>(
        manifest: &ExtractionManifest,
        request_id: Uuid,
        data: T,
    ) -> Result<Self, ContractError> {
        let data = serde_json::to_value(data)
            .map_err(|error| ContractError::new(format!("result data JSON 无效: {error}")))?;
        Ok(Self(WireProtocolEnvelope::Result {
            protocol_version: manifest.protocol_version,
            request_id,
            outcome: Outcome::Ok { data },
        }))
    }

    pub fn error(
        manifest: &ExtractionManifest,
        request_id: Uuid,
        code: ErrorCode,
        retryable: bool,
    ) -> Result<Self, ContractError> {
        if !manifest.error_codes.contains(&code) {
            return Err(ContractError::new("结果包含未知错误码"));
        }
        Ok(Self(WireProtocolEnvelope::Result {
            protocol_version: manifest.protocol_version,
            request_id,
            outcome: Outcome::Error {
                error: ErrorBody { code, retryable },
            },
        }))
    }
}

pub fn decode_protocol_envelope(
    value: Value,
    manifest: &ExtractionManifest,
) -> Result<ProtocolEnvelope, ContractError> {
    let envelope: WireProtocolEnvelope = serde_json::from_value(value)
        .map_err(|error| ContractError::new(format!("protocol envelope JSON 无效: {error}")))?;
    envelope.validate(manifest)?;
    Ok(ProtocolEnvelope(envelope))
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(tag = "status", rename_all = "lowercase", deny_unknown_fields)]
pub enum Outcome {
    Ok { data: Value },
    Error { error: ErrorBody },
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ErrorBody {
    pub code: ErrorCode,
    pub retryable: bool,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EnvelopeFixture {
    pub fixture_version: u32,
    pub cases: Vec<EnvelopeFixtureCase>,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct EnvelopeFixtureCase {
    pub name: String,
    pub route: String,
    pub accept: bool,
    pub value: Value,
}
